source/gameengine/GameLogic/Joystick/SCA_Joystick.cpp

```diff
@@ -74,95 +72,82 @@ SCA_Joystick::~SCA_Joystick()
 }
 
 SCA_Joystick *SCA_Joystick::m_instance[JOYINDEX_MAX];
-int SCA_Joystick::m_joynum = 0;
-int SCA_Joystick::m_refCount = 0;
+bool SCA_Joystick::m_joystickupdatestatus = false;
+
 
-SCA_Joystick *SCA_Joystick::GetInstance( short int joyindex )
+void SCA_Joystick::Init()
 {
-#ifndef WITH_SDL
-	return NULL;
-#else  /* WITH_SDL */
-	if (!SDL_CHECK(SDL_InitSubSystem)) {
-		return NULL;
-	}
-	if (joyindex < 0 || joyindex >= JOYINDEX_MAX) {
-		JOYSTICK_ECHO("Error-invalid joystick index: " << joyindex);
-		return NULL;
+#ifdef WITH_SDL
+
+	if (!(SDL_CHECK(SDL_InitSubSystem)) ||
+	    !(SDL_CHECK(SDL_GameControllerAddMapping)) ||
+	    !(SDL_CHECK(SDL_NumJoysticks))) {
+		return;
 	}
 
-	if (m_refCount == 0) 
-	{
-		int i;
-
-		/* The video subsystem is required for joystick input to work. However,
-		 * when GHOST is running under SDL, video is initialized elsewhere. We
-		 * also need to set the videodriver to dummy, and do it here to avoid
-		 * interfering with addons that may use SDL too.
-		 *
-		 * We also init SDL once only. */
-#  ifdef WITH_GHOST_SDL
-		int success = (SDL_InitSubSystem(SDL_INIT_JOYSTICK) != -1 );
-#  else
-		/* set and restore environment variable */
-		char *videodriver = getenv("SDL_VIDEODRIVER");
-		BLI_setenv("SDL_VIDEODRIVER", "dummy");
-
-#if SDL_VERSION_ATLEAST(2, 0, 0)
-		int success = (SDL_InitSubSystem(SDL_INIT_JOYSTICK) != -1 );
-#else
-		int success = (SDL_InitSubSystem(SDL_INIT_JOYSTICK | SDL_INIT_VIDEO) != -1 );
-#endif
+	/* Initializing Game Controller related subsystems */
+	bool success = (SDL_InitSubSystem(SDL_INIT_GAMECONTROLLER | SDL_INIT_HAPTIC) != -1 );
 
-		BLI_setenv("SDL_VIDEODRIVER", videodriver);
-#  endif
+	if (success) {
 
-		if (!success) {
-			JOYSTICK_ECHO("Error-Initializing-SDL: " << SDL_GetError());
-			return NULL;
-		}
-		
+		/* Loading Game Controller mapping data base from a string */
+		unsigned short i = 0;
+		const char *mapping_string = NULL;
+		mapping_string = controller_mappings[i];
+
+		while (mapping_string) {
+			SDL_GameControllerAddMapping(mapping_string);
+			i++;
+			mapping_string = controller_mappings[i];
+	    }
+#if 0
+		/* Creating Game Controllers that are already connected */
 		m_joynum = SDL_NumJoysticks();
-		
-		for (i = 0; i < m_joynum; i++) {
-			m_instance[i] = new SCA_Joystick(i);
-			m_instance[i]->CreateJoystickDevice();
+		printf("m_refCount = %i\n", m_refCount);
+		for (int j = 0; j < m_joynum; j++) {
+			m_instance[j] = new SCA_Joystick(j);
+			m_instance[j]->CreateJoystickDevice();
+			m_instancemapping[j] = ++m_refCount;
 		}
-		m_refCount = 1;
+#endif
 	}
-	else
-	{
-		m_refCount++;
+	else {
+		printf("Error initializing SDL Game Controller: %s\n", SDL_GetError());
 	}
-	return m_instance[joyindex];
-#endif /* WITH_SDL */
+#endif
 }
 
-void SCA_Joystick::ReleaseInstance()
+void SCA_Joystick::Close()
```

Comment:
If this function need to be called only one time and not per scene because there one joystick manager per scene, then move its call in `LA_Launcher::ExitEngine`. Same for `Init` in `LA_Launcher::InitEngine`